### ISSUE TITLE
Troubles in installation packages with pip3 install -r requirements.txt

### ISSUE THREAD

**Issue by lxlofpku** [stefmolin/Hands-On-Data-Analysis-with-Pandas-2nd-edition#24]
### Required attestation
- [Y] I have **completely and exactly** followed the virtual environment setup instructions from the book or [README](https://github.com/stefmolin/Hands-On-Data-Analysis-with-Pandas-2nd-edition#notes-on-environment-setup) in absence of the book.
- [Y] I have cloned either this repository or my fork of this repository so that I have all necessary files on my local machine.
- [Y ] I have checked that this issue has not already been reported or resolved [here](https://github.com/stefmolin/Hands-On-Data-Analysis-with-Pandas-2nd-edition/issues?q=is%3Aissue+label%3Atroubleshooting).
- [Y] I am aware that there is a [pre-built Binder environment](https://mybinder.org/v2/gh/stefmolin/binder-environments/2nd_edition?urlpath=git-pull%3Frepo%3Dhttps%253A%252F%252Fgithub.com%252Fstefmolin%252FHands-On-Data-Analysis-with-Pandas-2nd-edition%26urlpath%3Dlab%252Ftree%252FHands-On-Data-Analysis-with-Pandas-2nd-edition%252F%26branch%3Dmaster) that I can use, but I want to install locally on my machine instead.

---

##### 1. Which OS are you using?
Mac airbook M2

##### 2. Which Python version are you using?
Python3 from the current anaconda3

##### 3. Are you using `conda` or `venv`?
conda

---

### Commands run and their outputs
Please provide **all** of the commands you ran as well as the traceback:

<details>

```
conda create --name book_env python=3.8

conda env list

conda activate book_env

pip3 install -r requirements.txt

```

</details>

### Screenshots
Collecting git+https://github.com/stefmolin/login-attempt-simulator.git@2nd_edition (from -r requirements.txt (line 15))
  Cloning https://github.com/stefmolin/login-attempt-simulator.git (to revision 2nd_edition) to /private/var/folders/s9/zk8r4hnd1p9dh9tf9s0xytfw0000gn/T/pip-req-build-pb8bm614
  Running command git clone --filter=blob:none --quiet https://github.com/stefmolin/login-attempt-simulator.git /private/var/folders/s9/zk8r4hnd1p9dh9tf9s0xytfw0000gn/T/pip-req-build-pb8bm614
  fatal: unable to access 'https://github.com/stefmolin/login-attempt-simulator.git/': HTTP/2 stream 1 was not closed cleanly before end of the underlying stream
  error: subprocess-exited-with-error
  
  × git clone --filter=blob:none --quiet https://github.com/stefmolin/login-attempt-simulator.git /private/var/folders/s9/zk8r4hnd1p9dh9tf9s0xytfw0000gn/T/pip-req-build-pb8bm614 did not run successfully.
  │ exit code: 128
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× git clone --filter=blob:none --quiet https://github.com/stefmolin/login-attempt-simulator.git /private/var/folders/s9/zk8r4hnd1p9dh9tf9s0xytfw0000gn/T/pip-req-build-pb8bm614 did not run successfully.
│ exit code: 128
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

=================================
By the way, I did not clone the files since I am not formiliar with the clone tool. Instead, I downloaded the whole .zip file then unzip the contents into my local directory.

Any help will be appreciated.

Best regards,
Li-Xin Li

**Comment by stefmolin** [comment]
@lxlofpku - First, for M1 Macs and I imagine M2 as well, you will need to use Python 3.9 instead of 3.8, so you need to recreate the environment.

The issue you pasted above might be related to the git config on your machine and the use of HTTP/2 instead of HTTP/1.1 — if you still have issues after recreating your environment, try running the following:
```
git config --global http.version HTTP/1.1
pip3 install -r requirements.txt
```

**Comment by stefmolin** [comment]
Try deleting the environment and following the steps [here](https://github.com/stefmolin/Hands-On-Data-Analysis-with-Pandas-2nd-edition#m1-mac-users).

**Comment by lxlofpku** [comment]
@stefmolin 

I tried to install mamba by "conda install mamba -n base -c conda-forge" but I get the following:
==============================
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: - 
Found conflicts! Looking for incompatible packages.
This can take several minutes.  Press CTRL-C to abort.
Examining conflict for backcall nbformat appscript snowballstemmer jupyterlab_widgets anaconda executing \ c
========================================
then it keep "examing" for several hours without finishement. So, I did not success in taling the mamba...I even tried to reinstall the anaconda, but it still like the above when running "conda install mamba -n base -c conda-forge".

I tried to install a new version of git then repeat the process with conda, then when I run "pip3 install -r requirements.txt" I get
==================================
(book_env) lxl@LideMacBook-Air Hands-On-Data-Analysis-with-Pandas-2nd-edition % pip3 install -r requirements.txt
Collecting git+https://github.com/stefmolin/login-attempt-simulator.git@2nd_edition (from -r requirements.txt (line 15))
  Cloning https://github.com/stefmolin/login-attempt-simulator.git (to revision 2nd_edition) to /private/var/folders/s9/zk8r4hnd1p9dh9tf9s0xytfw0000gn/T/pip-req-build-gp3d_qhd
  Running command git clone --filter=blob:none --quiet https://github.com/stefmolin/login-attempt-simulator.git /private/var/folders/s9/zk8r4hnd1p9dh9tf9s0xytfw0000gn/T/pip-req-build-gp3d_qhd
  Running command git checkout -q 76e065a122343ebdd51c1cae98fbb8285411c0ff
  Resolved https://github.com/stefmolin/login-attempt-simulator.git to commit 76e065a122343ebdd51c1cae98fbb8285411c0ff
  Preparing metadata (setup.py) ... done
Collecting git+https://github.com/stefmolin/ml-utils.git@2nd_edition (from -r requirements.txt (line 16))
  Cloning https://github.com/stefmolin/ml-utils.git (to revision 2nd_edition) to /private/var/folders/s9/zk8r4hnd1p9dh9tf9s0xytfw0000gn/T/pip-req-build-9zs3shx6
  Running command git clone --filter=blob:none --quiet https://github.com/stefmolin/ml-utils.git /private/var/folders/s9/zk8r4hnd1p9dh9tf9s0xytfw0000gn/T/pip-req-build-9zs3shx6
  fatal: cannot access 'https://github.com/stefmolin/ml-utils.git/'：OpenSSL SSL_read: error:0A000126:SSL routines::unexpected eof while reading, errno 0
  error: subprocess-exited-with-error
  
  × git clone --filter=blob:none --quiet https://github.com/stefmolin/ml-utils.git /private/var/folders/s9/zk8r4hnd1p9dh9tf9s0xytfw0000gn/T/pip-req-build-9zs3shx6 did not run successfully.
  │ exit code: 128
  ╰─> See above for output.
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× git clone --filter=blob:none --quiet https://github.com/stefmolin/ml-utils.git /private/var/folders/s9/zk8r4hnd1p9dh9tf9s0xytfw0000gn/T/pip-req-build-9zs3shx6 did not run successfully.
│ exit code: 128
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.
======================================================

So far, no success.

Best regards,
Li-Xin

**Comment by stefmolin** [comment]
I often find that conda hangs like that, which was why I suggested mamba. It sounds like you may have some configuration issues on your machine. All issues have been related to making web requests (protocol, SSL) — you are indeed connected to the Internet when you try to install, correct?

Try deleting your environment and creating it this way:
```
conda env create --file environment.yml
```

If that doesn't work, the only other idea I have to install locally would be the following:
1. Recreate the conda environment with Python 3.9
2. Remove the last 4 lines of the requirements.txt file and then run the pip install command from before.
3.  Download the zip files for each of those repos, unzip, change into that directory, and then run `pip3 install .` (this exact command) from there: 
    - https://github.com/stefmolin/login-attempt-simulator
    - https://github.com/stefmolin/ml-utils
    -  https://github.com/stefmolin/stock-analysis
5. Change back to the book's directory and run `pip3 install ./visual-aids`

Note that doing this means that some of the code in the repos you download in step 3 may be different from what is presented in the book. The code in the book is under the tag `2nd_edition` in each repo.

On another note, [Binder](https://mybinder.org/v2/gh/stefmolin/Hands-On-Data-Analysis-with-Pandas-2nd-edition/master) is also a great option if you experience local installation issues, since it is already configured.

**Comment by lxlofpku** [comment]
@stefmolin 

Hi Stefanie:

I followed the "four steps" approach but encountered a problem again in the step 2. After removing the last four lines from requirments.txt, I ran "pip3 install -r requirements.txt" but got:

..............
building 'npymath' library
            compiling C sources
            C compiler: clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -fwrapv -O2 -Wall -fPIC -O2 -isystem /Users/lxl/opt/anaconda3/envs/book_env/include -arch x86_64 -I/Users/lxl/opt/anaconda3/envs/book_env/include -fPIC -O2 -isystem /Users/lxl/opt/anaconda3/envs/book_env/include -arch x86_64
      
            compile options: '-Ibuild/src.macosx-10.9-x86_64-3.9/numpy/core/src/npymath -Inumpy/core/include -Ibuild/src.macosx-10.9-x86_64-3.9/numpy/core/include/numpy -Inumpy/core/src/common -Inumpy/core/src -Inumpy/core -Inumpy/core/src/npymath -Inumpy/core/src/multiarray -Inumpy/core/src/umath -Inumpy/core/src/npysort -I/Users/lxl/opt/anaconda3/envs/book_env/include/python3.9 -Ibuild/src.macosx-10.9-x86_64-3.9/numpy/core/src/common -Ibuild/src.macosx-10.9-x86_64-3.9/numpy/core/src/npymath -Ibuild/src.macosx-10.9-x86_64-3.9/numpy/core/src/common -Ibuild/src.macosx-10.9-x86_64-3.9/numpy/core/src/npymath -c'
            llvm-ar: adding 4 object files to build/temp.macosx-10.9-x86_64-cpython-39/libnpymath.a
            error: Command "/var/folders/sy/f16zz6x50xz3113nwtb9bvq00000gp/T/abs_5a6dxej6x9/croot/python-split_1665695291976/_build_env/bin/llvm-ar rcs build/temp.macosx-10.9-x86_64-cpython-39/libnpymath.a build/temp.macosx-10.9-x86_64-cpython-39/numpy/core/src/npymath/npy_math.o build/temp.macosx-10.9-x86_64-cpython-39/build/src.macosx-10.9-x86_64-3.9/numpy/core/src/npymath/ieee754.o build/temp.macosx-10.9-x86_64-cpython-39/build/src.macosx-10.9-x86_64-3.9/numpy/core/src/npymath/npy_math_complex.o build/temp.macosx-10.9-x86_64-cpython-39/numpy/core/src/npymath/halffloat.o" failed with exit status 127
            [end of output]
      
        note: This error originates from a subprocess, and is likely not a problem with pip.
      error: legacy-install-failure
      
      × Encountered error while trying to install package.
      ╰─> numpy
      
      note: This is an issue with the package mentioned above, not pip.
      hint: See above for output from the failure.
      [end of output]
  
  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× pip subprocess to install build dependencies did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip.

So, it failed again. I am so sorry for bothering so much.

With best regards,
Li-Xin

**Comment by stefmolin** [comment]
The problem is clearly with the configuration on your machine, which I am not going to be able to solve for you. You will have to Google to figure out what system libraries you are missing, such that you can't even install packages. There seem to be some missing C dependencies and issues with openssl, from what we have seen so far.

Conda might be able to handle these for you if you remove the last 5 lines from the environment.yml file and then do this as suggested above:
>Try deleting your environment and creating it this way:
>conda env create --file environment.yml

After that you would continue from step 3 above.

If you are not comfortable on the command line (which I'm inferring since you didn't want to use `git clone`), then **I strongly recommend you use [Binder](https://mybinder.org/v2/gh/stefmolin/Hands-On-Data-Analysis-with-Pandas-2nd-edition/master)**, since fixing your local configuration will require a lot of Googling and using the command line to install missing system dependencies.

**Comment by lxlofpku** [comment]
@stefmolin 
Hi Stefanie:

This time I succeeded, I think. It appears that scikit-learn0.23.2 is not supported by Python3.9. I generated the book_env with Python3.8, then all 1-4 stepps went through successfully.

Best regards,
Li-Xin

**Comment by stefmolin** [comment]
I just did a fresh install on an M1 Mac and version 0.23.2 of scikit-learn is available for Python 3.9 if you use conda. Often with older versions, conda is your best bet since it will handle the system dependencies that you may be missing. If you want to use pip, you have to hunt down what is missing. Anyway, I'm glad you got it to work.
